core/storage/translation/gae_models.py

```diff
@@ -0,0 +1,147 @@
+        required=True, indexed=True)
+    # The language code of the untranslated text.
+    source_language_code = datastore_services.StringProperty(
+        required=True, indexed=True)
+    # The language code of the translation.
+    target_language_code = datastore_services.StringProperty(
+        required=True, indexed=True)
+    # The translation.
+    translated_text = datastore_services.TextProperty(
+        required=True, indexed=False)
+
+    @classmethod
+    def create(
+            cls, source_language_code, target_language_code, origin_text,
+            translated_text):
+        """Creates a new MachineTranslatedTextModel instance and returns its
+        ID.
+
+        Args:
+            source_language_code: str. The language code of the untranslated
+                text.
+            target_language_code: str. The language code of the translation.
+            origin_text: str. The untranslated text.
+            translated_text: str. The translation.
+
+        Returns:
+            str. The id of the newly created MachineTranslatedTextInstance.
+        """
+        hashed_origin_text = utils.convert_to_hash(origin_text, 50)
+        entity_id = cls._generate_id(
+            source_language_code, target_language_code, hashed_origin_text)
+        translation_entity = cls(
+            id=entity_id,
+            hashed_origin_text=hashed_origin_text,
+            source_language_code=source_language_code,
+            target_language_code=target_language_code,
+            origin_text=origin_text,
+            translated_text=translated_text)
+        translation_entity.put()
+        return entity_id
+
+    @classmethod
+    def _generate_id(
+            cls, source_language_code, target_language_code,
+            hashed_origin_text):
+        """Generates a valid key for a MachineTranslatedTextModel.
+
+        Args:
+            source_language_code: str. The language code of the untranslated
+                text.
+            target_language_code: str. The language code of the translation.
+            hashed_origin_text: str. An sha-1 hash of the origin_text.
+
+        Returns:
+            str. The generated ID for this entity of the form
+
+            [source_language_code]:[target_language_code]:[hashed_origin_text].
+        """
+
+        return (
+            '%s:%s:%s' % (
+                source_language_code, target_language_code, hashed_origin_text)
+        )
+
+    @classmethod
+    def get_translation_for_text(
+            cls, source_language_code, target_language_code, origin_text):
+        """Gets MachineTranslatedTextModel by language codes and origin text.
+        Returns None if no translation exists for the given parameters.
+
+        Args:
+            source_language_code: str. The language of the origin_text.
+            target_language_code: str. The language being translated to.
+            origin_text: str. The text to be translated.
```

Comment:
Slight nit, but maybe instead use: ``The text to be or that was translated.`` (since in this context we may be fetching something already translated).

Reply:
Though, this actually begs a different question: will we always know the translated text upon creation? If so, then perhaps ``The text that was, or may need to be, translated.`` would actually be clearer.

Reply:
I updated the language in the docstrings to be more descriptive, PTAL.